lib/utils/property-references.js

```diff
@@ -92,7 +92,10 @@ module.exports = {
 /**
  * @param {RuleContext} context The rule context.
  */
-function definePropertyReferenceExtractor(context) {
+function definePropertyReferenceExtractor(
+  context,
+  { stopReportingUnknown = true, stopReportingReturn = true } = {}
```

Comment:
Your changes are almost identical to what I had in mind. Thank you.

However, I think the option names are a bit confusing for this function. (It's not about the option names of the rule.)

For example, how about naming it like this:

```suggestion
  { unknownMemberAsUnreferenced = false, returnStatementAsUnreferenced = false } = {}
```

(Note that I set the initial value to false.)

Reply:
First of all, great that we're aligned :). In terms of changing the names, I still find the actual config options a bit confusing (I don't like 'negatives' in the options, so it would be good to remove the 'stop' part for example). Perhaps we can work out a name that will work in both places?

Since both are specifically about access to `this`, it seems like a good idea to get that into the option. `this[variable]` is a dynamic access to `this`, and `return this` is a potential access via the returned value, so could we use `reportDynamicThis` and `reportReturnedThis` as both the config and argument names? That has the advantage of not needing to set anything for the default options (to avoid changing the current behaviour)—it would only change by _adding_ those options. It also means the tests for this PR can be simplified, as no new default options are being introduced, and only explicit tests for the new functionality are needed.

I will update with those changes—easy to revert if you have a strong argument against changing the names...

Reply:
Hmm... Sorry but the naming is still confusing to me.
I think the `report` is confusing because the `definePropertyReferenceExtractor` function does not call `context.report()`. I think it should mean whether all properties are considered referenced or all properties are considered unreferenced.
Also, I think `This` is confusing because there are cases other than `this`. We may be better to use `Expression` instead.

Reply:
Yes, that makes sense. So the lower level utility is concerned with if things are considered referenced or not, whereas the outer rule is concerned with what to report—I will rename those accordingly.

> Also, I think This is confusing because there are cases other than this. We may be better to use Expression instead.

Since my use case is specifically around `this`, it seems like there should be some other cases to consider then. Can you share any examples which could be used in the tests, where this option should apply?

Reply:
>  Can you share any examples which could be used in the tests

For example, the following test does not return `this`.

```js
    {
      filename: 'test.vue',
      code: `
      <script>
        export default {
          data () {
            return {
              foo: {
                bar: 1
              },
              baz: 2
            }
          },
          methods: {
            handler () {
              const vm = this
              console.log(vm.baz)
              return vm.foo
            },
          }
        }
      </script>
      `,
      options: [{
        groups: ['data'],
        deepData: true,
        stopReporting: []
      }],
      errors: [
        {
          message: "'foo.bar' of data found, but never used.",
          line: 7,
        }
      ]
    }
```

Reply:
Added this in tests